File: LeetCode/DPMultiDimension/python/maximalSquare.py
from typing import List


class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:

        ROWS = len(matrix)
        COLS = len(matrix[0])
        cache = {} # map each(r,c) -> maxLength of square

        def helper(r,c):
            if r >= ROWS or c >= COLS:
                return 0
            
            if (r,c) not in cache:
                down = helper(r + 1, c)
                right = helper(r, c + 1)
                diag = helper(r + 1, c + 1)

                cache[(r,c)] = 0
                if matrix[r][c]  == "1":
                    cache[(r,c)] = 1 + min(down, right, diag)
            
            return cache[(r,c)]
        
        helper(0,0)
        return max(cache.values()) ** 2
    
    def maximalSquareDP(self, matrix: List[List[str]]) -> int:
        if not matrix:
            return 0
        rows, cols = len(matrix), len(matrix[0])
        dp = [[0] * (cols + 1) for _ in range(rows + 1)] 
        max_side = 0
        for i in range(1, rows + 1):
            for j in range(1, cols + 1):
                if matrix[i - 1][j - 1] == '1':
                    dp[i][j] = min(dp[i - 1][j], dp[i][j - 1],dp[i - 1][j - 1]) + 1
                    max_side = max(max_side, dp[i][j])

        return max_side * max_side

if __name__ == "__main__":
    obj = Solution()

    input1 = {"matrix": [["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]}
    expected1 = 4
    result1 = obj.maximalSquareDP(input1["matrix"])
    print("Input: Matrix: {}".format(input1["matrix"]))
    print("Expected: {}".format(expected1))
    print("Result: {}".format(result1))

    input2 = {"matrix": [["0","1"],["1","0"]]}
    expected2 = 1
    result2 = obj.maximalSquareDP(input2["matrix"])
    print("Input: Matrix: {}".format(input2["matrix"]))
    print("Expected: {}".format(expected2))
    print("Result: {}".format(result2))

    input3 = {"matrix": [["0"]]}
    expected3 = 0
    result3 = obj.maximalSquareDP(input3["matrix"])
    print("Input: Matrix: {}".format(input3["matrix"]))
    print("Expected: {}".format(expected3))
    print("Result: {}".format(result3))

File: LeetCode/DPMultiDimension/JavaScript/maximalSquare.js
// Given an m x n binary matrix filled with 0's and 1's, find the largest square containing only 1's and return its area.

// Example 1:

// Input: matrix = [["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]
// Output: 4
// Example 2:

// Input: matrix = [["0","1"],["1","0"]]
// Output: 1
// Example 3:

// Input: matrix = [["0"]]
// Output: 0

const Result = { PASS: "\x1b[92mPASS\x1b[0m", FAIL: "\x1b[91mFAIL\x1b[0m" };

class MaximalSquareRecord {
  constructor(matrix, expected) {
    this.matrix = matrix;
    this.expected = expected;
  }
}

class Solution {
  /**
   * Top-down memoization: each (r,c) caches the side length of the largest
   * square whose top-left corner is that cell.
   * @param {character[][]} matrix
   * @return {number}
   */
  maximalSquare(matrix) {
    let rows = matrix.length;
    let cols = matrix[0].length;
    let cache = {}; // map each(r,c) -> maxLength of square

    let helper = (r, c) => {
      if (r >= rows || c >= cols) {
        return 0;
      }

      if (!(`${r},${c}` in cache)) {
        let down = helper(r + 1, c);
        let right = helper(r, c + 1);
        let diag = helper(r + 1, c + 1);

        cache[`${r},${c}`] = 0;
        if (matrix[r][c] == "1") {
          cache[`${r},${c}`] = 1 + Math.min(down, right, diag);
        }
      }
      return cache[`${r},${c}`];
    };

    helper(0, 0);

    // Find the maxvalue from the array of values
    let res = [...Object.values(cache)].reduce((a, b) => (a > b ? a : b));

    return Math.pow(res, 2);
  }

  /**
   * Bottom-up tabulation: dp[i][j] is the side length of the largest square
   * whose bottom-right corner is matrix[i-1][j-1], padded with a zero row/col.
   * @param {character[][]} matrix
   * @return {number}
   */
  maximalSquareDP(matrix) {
    if (!matrix) {
      return 0;
    }

    let rows = matrix.length;
    let cols = matrix[0].length;

    let dp = Array(rows + 1)
      .fill()
      .map(() => Array(cols + 1).fill(0));

    let max_side = 0;

    for (let i = 1; i < rows + 1; i += 1) {
      for (let j = 1; j < cols + 1; j += 1) {
        if (matrix[i - 1][j - 1] === "1") {
          dp[i][j] = 1 + Math.min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]);
          max_side = Math.max(max_side, dp[i][j]);
        }
      }
    }

    return Math.pow(max_side, 2);
  }
}

function testSolution(record) {
  const solution = new Solution();
  const result = solution.maximalSquareDP(record.matrix);
  const pass = result === record.expected;

  console.log(`Input: matrix = ${JSON.stringify(record.matrix)}`);
  console.log(`Expected: ${record.expected}`);
  console.log(`Result: ${result}`);
  console.log(pass ? Result.PASS : Result.FAIL);
}

const records = [
  new MaximalSquareRecord(
    [
      ["1", "0", "1", "0", "0"],
      ["1", "0", "1", "1", "1"],
      ["1", "1", "1", "1", "1"],
      ["1", "0", "0", "1", "0"],
    ],
    4
  ),
  new MaximalSquareRecord(
    [
      ["0", "1"],
      ["1", "0"],
    ],
    1
  ),
  new MaximalSquareRecord([["0"]], 0),
  new MaximalSquareRecord([["1"]], 1),
  new MaximalSquareRecord(
    [
      ["1", "1", "1"],
      ["1", "1", "1"],
      ["1", "1", "1"],
    ],
    9
  ),
];

records.forEach((record, index) => {
  console.log(`# Test case ${index + 1}`);
  testSolution(record);
  console.log("----------------------------------------");
});
